### PR TITLE
Feature/weapon effect tooltips

### LINKED ISSUE #121: Add the percentage property to rune tooltips
Currently, there is no way to add the value with a percentage to our custom rune tooltip. The ability specials should have an optional property that is also sent to panorama with the values. The panorama will then localize with a percentage when relevant.

Opened from Discord by **Shush#1242**
Follow the conversation [here](https://discord.com/channels/501306949434867713/826466764606799882/849402505082306570)

### LINKED ISSUE #141: Create keyword tooltip
Create a portion in the runes tooltips below the rune's description, where all keywords in the text, such as generic rune effects, are shown while the player holds ALT.

Opened from Discord by **Shush#1242**
Follow the conversation [here](https://discord.com/channels/501306949434867713/826466764606799882/850740168553922581)

### PR DESCRIPTION
**Pull request by Nibuja05**
Fixed #141 

Automatically marks all generic weapon effects in the description of any rune and displays an info box for it:
- to use this feature, simply use the `GenericWeaponEffectName` enum in the localization of any rune or generic effect description
- alternatively you can directly use: `%burning%`, `%zapping%` etc

- info box is only displayed while alt is pressed

Fixes #121 

Adds special rune value localization:
-  all special rune values not defined in the localization will not be displayed
- adding `%` as prefix marks the value as percentage
- adding `!` as prefix marks the value as duration
- special values can also be used inside descriptions: simple wrap them in `%`
e.g. `%damage%`, `%interval%`.

### CODE REVIEW

**Review by Shushishtok — CHANGES_REQUESTED** [review]
Really good work! Some suggestions to make it super perfect :D

File: common/weapons.d.ts [hunk 1]
```diff
@@ -120,3 +120,12 @@ declare const enum WeaponSoundEventName {
 	FireballCast = "Fireball.Cast",
 	FireballImpact = "Fireball.Impact",
 }
+
+declare const enum GenericWeaponEffectName {
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply common/weapons.d.ts:124]: Done.

File: content/panorama/scripts/custom_game/util.ts [hunk 2]
```diff
@@ -195,3 +195,87 @@ function GetCurrentGold(): number {
...
+	let text = LocalizeGenericWeaponEffectName(effect);
+	return MarkText(text);
+}
+
+/**
+ * Localize the generic weapon effect name.
+ * @param effect effect to localize
+ * @returns localized name
+ */
+function LocalizeGenericWeaponEffectName(effect: GenericWeaponEffectName): string {
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/util.ts:214]: Done.

File: content/panorama/scripts/custom_game/util.ts [hunk 3]
```diff
@@ -195,3 +195,87 @@ function GetCurrentGold(): number {
...
+ * Periodically checks if the alt button status has changed.
+ * If yes, registered callbacks may be called.
+ */
+function CheckAltOption() {
+	if (GameUI.IsAltDown() && altDown == false) {
+		altDown = true;
+		for (const callback of altCallbacks) {
+			callback(altDown);
+		}
+	} else if (!GameUI.IsAltDown() && altDown == true) {
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/util.ts:252]: Done.

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 4]
```diff
@@ -506,7 +559,87 @@ abstract class PanoramaWeaponBase {
...
+		GenericWeaponEffectName.CONFUSION,
+		GenericWeaponEffectName.DRENCHED,
+		GenericWeaponEffectName.FREEZING,
+		GenericWeaponEffectName.FROZEN,
+		GenericWeaponEffectName.ZAPPED,
+	];
+	let foundEffects: GenericWeaponEffectName[] = [];
+	let newText = text;
+	for (const effect of effects) {
+		if (newText.indexOf(effect) >= 0) {
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/weapon_base.ts:584]: Done.

File: content/panorama/scripts/custom_game/weapon_base.ts [hunk 5]
```diff
@@ -506,7 +559,87 @@ abstract class PanoramaWeaponBase {
...
+ * @param values values to replace the found special values with
+ * @param weaponName name of the current weapon (needed for localization)
+ * @returns text with replaced special values
+ */
+function MarkWeaponSpecialValues(text: string, values: CurrentWeaponValue[], weaponName: string): string {
+	if (text.search(/\%\w+\%/) < 0) {
+		return text;
+	}
+	for (const value of Object.values(values)) {
+		if (text.indexOf(`%${value.name}%`) >= 0) {
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/weapon_base.ts:632]: Done.

File: content/panorama/styles/custom_game/weapon_base.css [hunk 6]
```diff
@@ -270,3 +272,42 @@
...
+.GenericEffectBody {
+	background-color: gradient(linear, 0% 0%, 0% 100%, from(#353d44), to(#2c343a));
+	color: #95a5b9;
+	text-shadow: 2px 2px 0px 1 #00000090;
+	font-size: 16px;
+	height: fit-children;
+	width: 100%;
+	padding: 8px;
+	padding-top: 6px;
+	/* margin-bottom: 6px; */
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply content/panorama/styles/custom_game/weapon_base.css:312]: Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done. Done.

File: content/panorama/styles/custom_game/weapon_base.css [hunk 7]
```diff
@@ -111,7 +112,8 @@
 	height: 15px;
 	horizontal-align: right;
 	border-radius: 5px;
-	background-color: rgb(65, 65, 65);
+	/* background-color: rgb(65, 65, 65); */
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply content/panorama/styles/custom_game/weapon_base.css:115]: Done.

File: content/panorama/scripts/custom_game/util.ts [hunk 8]
```diff
@@ -195,3 +195,87 @@ function GetCurrentGold(): number {
...
+
+let altDown = false;
+let altRunning: ScheduleID | undefined;
+let altCallbacks: Set<(down: boolean) => void> = new Set();
+/**
+ * Periodically checks if the alt button status has changed.
+ * If yes, registered callbacks may be called.
+ */
+function CheckAltOption() {
+	if (GameUI.IsAltDown() && altDown == false) {
```
[the remark this answers is not in the document]
Comment (Nibuja05) [reply content/panorama/scripts/custom_game/util.ts:247]: Done.

**Review by Shushishtok — APPROVED** [review]

**Pull request merged**